### PR TITLE
[AR-3264] RabbitStew catches SIGTERM and SIGINT and sends SignalExcep…

### PR DESCRIPTION
**Pull request by codyplungis**
…tion to workers

When sending SIGTERM or SIGINT(Ctrl+c) to a daemon it would just kill the daemon and the worker threads would not get a chance to clean up. Now RabbitStew will trap the signal and send an exception to the workers allowing them die gracefully.

https://indigobio.atlassian.net/browse/AR-3264

### CODE REVIEW

File: lib/rabbit_stew/service/runner.rb [hunk 1]
```diff
@@ -5,6 +5,9 @@
 module RabbitStew
   module Service
     class Runner
+      # The container will wait a default of 5 seconds when stopping, so we want less than that
```
Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: What's the downside to having the container shutdown timeout? It's harsher?

File: lib/rabbit_stew/service/runner.rb [hunk 2]
```diff
@@ -49,6 +54,20 @@ def disable_output_buffering
         # Necessary for docker and integration tests
         $stdout.sync = $stderr.sync = true
       end
+
+      def handle_signal(signal)
+        Signal.trap(signal) do
```
Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: Don't try to stop the subscriber again if a second signal comes in while we're shutting down. Something like:

```ruby
Signal.trap(signal) do
  unless @shutting_down
    @shutting_down = true
    @subscriber.stop do ...
    ...
  end
end
```

You can avoid the instance var (to make the signal handling more self-contained) if you rearrange things a little:

```ruby
def handle_signals(*signals)
  shutting_down = false
  signals.each do |signal|
    Signal.trap(signal) do
      unless shutting_down
        shutting_down = true
        @subscriber.stop do ...
        ...
      end
    end
  end
end

File: lib/rabbit_stew/service/runner.rb [hunk 3]
```diff
@@ -5,6 +5,9 @@
 module RabbitStew
   module Service
     class Runner
+      # The container will wait a default of 5 seconds when stopping, so we want less than that
+      # https://github.com/phusion/baseimage-docker/blob/master/image/bin/my_init
+      WORKER_JOIN_TIMEOUT = 4
```
Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: units? `WORKER_JOIN_TIMEOUT_SECONDS`

Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: Instead of hardcoding the `4`, should this be based on the same `KILL_PROCESS_TIMEOUT` env var the base image init file is reading? (Is the env var exposed to the daemon process?)

Comment (codyplungis) [inline lib/rabbit_stew/service/runner.rb]: It appears we don't actually set that variable when running in production. I couldn't find it in prod's [all.env](https://github.com/indigobio/prod-env/blob/master/environments/prod/all.env).

File: spec/service/runner_spec.rb [hunk 4]
```diff
@@ -98,7 +111,82 @@
...
+      end
+    end
+    sleep 1
+    pid
+  end
+
+  def worker_proc(exception_occurred, no_exception, pipe_write)
+    proc do |msg|
+      begin
+        sleep 5
```
Comment (wintonpc) [inline spec/service/runner_spec.rb]: `develop` tests take 1.5 seconds to run. `b-terminateWorkers` tests take 14 seconds to run. Fix.

File: lib/rabbit_stew/service/runner.rb (hunk 3 above)
Comment (jwliechty) [inline lib/rabbit_stew/service/runner.rb]: If this is to be something configurable, then we should pass it in from the top instead of reading environment variable directly.

I would be in more of a favor of just setting it high, like 30 seconds.

File: spec/service/runner_spec.rb [hunk 5]
```diff
@@ -98,7 +111,82 @@
...
+    end
+  end
+
+  def start_daemon_with_worker(exception_occurred, no_exception, pipe_write)
+    pid = Process.fork do
+      start(update_sites: false) do
+        subject.worker(source, &worker_proc(exception_occurred, no_exception, pipe_write))
+      end
+    end
+    sleep 1
```
Comment (wintonpc) [inline spec/service/runner_spec.rb]: It's more robust (and efficient) to use a signalling technique to communicate that the child has started up.

File: lib/rabbit_stew/service/runner.rb [hunk 6]
```diff
@@ -49,6 +54,20 @@ def disable_output_buffering
         # Necessary for docker and integration tests
         $stdout.sync = $stderr.sync = true
       end
+
+      def handle_signal(signal)
+        Signal.trap(signal) do
+          # Stop rabbit subscribers so no new workers are created
+          @subscriber.stop do
```
Comment (jwliechty) [inline lib/rabbit_stew/service/runner.rb]: Maybe put a log message above this line indicating shutting down (or which signal was just received). That is typical of applications of which a signal is sent where there might be some time before handling completes.

File: spec/service/runner_spec.rb [hunk 7]
```diff
@@ -1,20 +1,33 @@
+require 'mercury/monadic'
+require 'mercury/test_utils'
 require 'rabbit_stew/service/runner'
 
+def expect_worker_exception(exception_occurred, pipe_read, pipe_write)
```
Comment (jwliechty) [inline spec/service/runner_spec.rb]: Place method at bottom within `Describe` block

File: spec/service/runner_spec.rb (hunk 5 above)
Comment (wintonpc) [inline spec/service/runner_spec.rb]: ... if that is in fact necessary. Why is this necessary?

Comment (wintonpc) [inline spec/service/runner_spec.rb]: I think I see now. Don't use sleeps. You have the pipe, so you can know exactly when the child changes states. For instance, you don't have to sleep between publishing the message and killing the child; have the child send "got msg" over the pipe and wait for that.

File: spec/service/runner_spec.rb (hunk 4 above)
Comment (jwliechty) [inline spec/service/runner_spec.rb]: I'm not sure how else you could test the timeout unless you wait the full amount. Maybe `Timecop` could help?

Comment (wintonpc) [inline spec/service/runner_spec.rb]: @jwliechty Make the timeout a parameter rather than hardcoding.

Comment (codyplungis) [inline spec/service/runner_spec.rb]: This sleep is only in place to ensure that between the worker does not write out the error message before we have a chance to kill the daemon. Whether this sleep is longer or shorter does not affect test run time except that if it's too short the test fails.

File: lib/rabbit_stew/service/runner.rb (hunk 1 above)
Comment (codyplungis) [inline lib/rabbit_stew/service/runner.rb]: We needed a timeout for the `join` otherwise we get exceptions saying the thread is sleeping. Since we need a timeout, we figured we should pick something reasonable. We could have just left it as the magic number `1`.

File: spec/service/runner_spec.rb (hunk 5 above)
Comment (codyplungis) [inline spec/service/runner_spec.rb]: This sleep is necessary because we need to ensure the daemon/worker has time to come up before we send the message. I have dropped this sleep to 0.1 seconds. If we sleep less then 0.1 seconds a worker never gets a message so the test fails.

Comment (codyplungis) [inline spec/service/runner_spec.rb]: It would be preferable to have some way of knowing the worker is ready to receive a message before sending it, but I have no idea how to do that.

File: lib/rabbit_stew/service/runner.rb (hunk 6 above)
Comment (codyplungis) [inline lib/rabbit_stew/service/runner.rb]: Added log message.

File: lib/rabbit_stew/service/runner.rb (hunk 3 above)
Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: @codyplungis oh I see now

File: spec/service/runner_spec.rb (hunk 5 above)
Comment (wintonpc) [inline spec/service/runner_spec.rb]: @codyplungis Then that's a missing feature of rabbit_stew that we need to implement: some kind of hook to let interested parties know when the workers are accepting messages. (Really, when the queues and exchanges have been declared.) Surely the rabbit_stew tests must deal with the same problem.

And in this case, you'd use the hook to send something over the pipe to let the parent know it's safe to send a message.

**Comment by codyplungis** [comment]
retest this please

**Comment by codyplungis** [comment]
retest this please.

**Comment by codyplungis** [comment]
retest this please

File: lib/rabbit_stew/service/runner.rb [hunk 8]
```diff
@@ -5,11 +5,14 @@
 module RabbitStew
   module Service
     class Runner
+      # The container will wait a default of 5 seconds when stopping, so we want less than that
+      # https://github.com/phusion/baseimage-docker/blob/master/image/bin/my_init
+      WORKER_JOIN_TIMEOUT_SECONDS = 4
 
-      def start(name, log_dir:, update_sites: true, &block)
+      def start(name, log_dir:, update_sites: true, when_started: nil, &block)
```
Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: initialize `when_started` to `proc{}` to avoid the need for nil checks later

File: lib/rabbit_stew/service/runner.rb [hunk 9]
```diff
@@ -49,6 +58,31 @@ def disable_output_buffering
...
+            unless shutting_down
+              shutting_down = true
+              shutdown(signal)
+            end
+          end
+        end
+      end
+
+      def shutdown(signal)
+        puts "Signal #{signal} received. Sending exceptions to workers then shutting down."
```
Comment (wintonpc) [inline lib/rabbit_stew/service/runner.rb]: Should this be logatron? Or is logging unreliable at this point?

File: spec/service/runner_spec.rb [hunk 10]
```diff
@@ -98,7 +106,90 @@
...
+        kill_daemon(pid, signal)
+        expect_worker_exception(pipe_read, pipe_write)
+      end
+    end
+  end
+
+  def start_daemon_with_worker(pipe_write)
+    pid = Process.fork do
+      hook = proc do
+        pipe_write.puts('started_listening')
```
Comment (wintonpc) [inline spec/service/runner_spec.rb]: use the `let`ed `workers_listening` instead

File: spec/service/runner_spec.rb [hunk 11]
```diff
@@ -98,7 +106,90 @@
...
+        expect_workers_get_signal_exception('INT')
+      end
+    end
+  end
+
+  def expect_workers_get_signal_exception(signal)
+    IO.pipe do |pipe_read, pipe_write|
+      run_spec do
+        pid = start_daemon_with_worker(pipe_write)
+        loop { break if pipe_read.gets == "#{workers_listening}\n" }
```
Comment (wintonpc) [inline spec/service/runner_spec.rb]: Shouldn't the first line read always be `workers_listening`? And if not, don't we want to know about it immediately? (fail)

Comment (wintonpc) [inline spec/service/runner_spec.rb]: This is one aspect of the test that I think we _do_ want to be brittle.

File: lib/rabbit_stew/service/runner.rb (hunk 9 above)
Comment (codyplungis) [inline lib/rabbit_stew/service/runner.rb]: If you try to log then you get an error about being unable to log in trap context.

File: spec/service/runner_spec.rb (hunk 11 above)
Comment (codyplungis) [inline spec/service/runner_spec.rb]: You are correct. I changed this to and expect.
